<!DOCTYPE HTML>
<html class="no-js" lang="zh-CN">
<head>
    <!--[if lte IE 9]>
<meta http-equiv="refresh" content="0;url=https://landau1994.github.io/warn.html">
<![endif]-->
<meta charset="utf-8">
<meta http-equiv="X-DNS-Prefetch-Control" content="on">
<link rel="dns-prefetch" href="https://landau1994.github.io">
<link rel="dns-prefetch" href="//www.google-analytics.com">
<link rel="prefetch" href="https://landau1994.github.io">
<link rel="prefetch" href="//www.google-analytics.com">


<link rel="prerender" href="https://landau1994.github.io">

<meta http-equiv="X-UA-Compatible" content="IE=Edge">
<meta name="renderer" content="webkit">
<meta name="viewport" content="width=device-width, initial-scale=1.0,user-scalable=no">
<meta http-equiv="mobile-agent" content="format=html5; url=https://landau1994.github.io">
<meta name="author" content="Longteng Wang">

<link rel="stylesheet" href="/css/JSimple.css">


<link rel="shortcut icon" href="/images/favicon.png">


<title>learnseurat_cite_seq - Coding for Life Sciences</title>

<meta name="keywords" content="[object Object]">

<meta name="description " content="关注从数据和模型中获取生物知识的博客">

    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({
            tex2jax: {
                inlineMath: [ ['$','$'], ["\\(","\\)"] ],
                processEscapes: true
            }
        });
    </script>


    

    

<meta name="generator" content="Hexo 5.2.0"><style>mjx-container[jax="SVG"] {
  direction: ltr;
}

mjx-container[jax="SVG"] > svg {
  overflow: visible;
}

mjx-container[jax="SVG"] > svg a {
  fill: blue;
  stroke: blue;
}

mjx-container[jax="SVG"][display="true"] {
  display: block;
  text-align: center;
  margin: 1em 0;
}

mjx-container[jax="SVG"][justify="left"] {
  text-align: left;
}

mjx-container[jax="SVG"][justify="right"] {
  text-align: right;
}

g[data-mml-node="merror"] > g {
  fill: red;
  stroke: red;
}

g[data-mml-node="merror"] > rect[data-background] {
  fill: yellow;
  stroke: none;
}

g[data-mml-node="mtable"] > line[data-line] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > rect[data-frame] {
  stroke-width: 70px;
  fill: none;
}

g[data-mml-node="mtable"] > .mjx-dashed {
  stroke-dasharray: 140;
}

g[data-mml-node="mtable"] > .mjx-dotted {
  stroke-linecap: round;
  stroke-dasharray: 0,140;
}

g[data-mml-node="mtable"] > svg {
  overflow: visible;
}

[jax="SVG"] mjx-tool {
  display: inline-block;
  position: relative;
  width: 0;
  height: 0;
}

[jax="SVG"] mjx-tool > mjx-tip {
  position: absolute;
  top: 0;
  left: 0;
}

mjx-tool > mjx-tip {
  display: inline-block;
  padding: .2em;
  border: 1px solid #888;
  font-size: 70%;
  background-color: #F8F8F8;
  color: black;
  box-shadow: 2px 2px 5px #AAAAAA;
}

g[data-mml-node="maction"][data-toggle] {
  cursor: pointer;
}

mjx-status {
  display: block;
  position: fixed;
  left: 1em;
  bottom: 1em;
  min-width: 25%;
  padding: .2em .4em;
  border: 1px solid #888;
  font-size: 90%;
  background-color: #F8F8F8;
  color: black;
}

foreignObject[data-mjx-xml] {
  font-family: initial;
  line-height: normal;
  overflow: visible;
}

.MathJax path {
  stroke-width: 3;
}

mjx-container {
  overflow: auto hidden;
}

mjx-container + br {
  display: none;
}
</style></head>
<body>
<div id="nav">
    <nav class="nav-menu">
        <a class="site-name current" href="/" title="CLS">CLS</a>
        <a class="site-index current" href="/"><i class="fa fa-home"></i><span>首页</span></a>
        <a href="/archives" title="归档"><i class="fa fa-archives"></i><span>归档</span></a>
        <a href="/tags" title="标签"><i class="fa fa-tags"></i><span>标签</span></a>
        <!-- custom single page of menus -->
        
        
        <a href="/help" title="帮助">
            <i class="fa fa-question-circle"></i>
            <span>帮助</span>
        </a>
        
    </nav>
</div>

<div class="nav-user">
    <a class="btn-search" href="#"><i class="fa fa-search"></i></a>
    <a class="btn-read-mode" href="#"><i class="fa fa-sun-o"></i></a>
    <a class="btn-sns-qr" href="javascript:"><i class="fa fa-telegram"></i></a>
</div>

<div id="wrapper" class="clearfix">
    <div id="body">
        <div class="main" id="main">
            <div id="cover">
    <div class="cover-img"></div>
    <div class="cover-info">
        
        <h1 class="cover-siteName">Coding for Life Science</h1>
        <h3 class="cover-siteTitle">Coding for Life Science</h3>
        <p class="cover-siteDesc">关注从数据和模型中获取生物知识的博客</p>
        <div class="cover-sns">
            
    &nbsp;&nbsp;<div class="btn btn-telegram">
        <a href="" target="_blank" title="telegram" ref="friend">
            <i class="fa fa-telegram"></i>
        </a>
    </div>

    &nbsp;&nbsp;<div class="btn btn-instagram">
        <a href="" target="_blank" title="instagram" ref="friend">
            <i class="fa fa-instagram"></i>
        </a>
    </div>

    &nbsp;&nbsp;<div class="btn btn-twitter">
        <a href="https://twitter.com/Landau1994" target="_blank" title="twitter" ref="friend">
            <i class="fa fa-twitter"></i>
        </a>
    </div>

    &nbsp;&nbsp;<div class="btn btn-github">
        <a href="https://github.com/Landau1994" target="_blank" title="github" ref="friend">
            <i class="fa fa-github"></i>
        </a>
    </div>


        </div>
    </div>
</div>

            <div class="page-title">
    <ul>
        <li><a href="/">最近</a></li>
        
            
                <li class="">
                    <a href="/categories/genomics" data-name="基因组学">基因组学</a>
                </li>
            
                <li class="">
                    <a href="/categories/implementation" data-name="编程实践">编程实践</a>
                </li>
            
                <li class="">
                    <a href="/categories/reference" data-name="文献阅读">文献阅读</a>
                </li>
            
                <li class="">
                    <a href="/categories/math" data-name="数学学习">数学学习</a>
                </li>
            
                <li class="">
                    <a href="/categories/others" data-name="其他">其他</a>
                </li>
            
        
        <li class="page-search">
    <form id="search" class="search-form">
        <input type="text"
               readonly="readonly"
               id="local-search-input-tip"
               placeholder="读物检索~" />
        <button type="button" disabled="disabled" class="search-form-submit"><i class="fa fa-search"></i></button>
    </form>
</li>

    </ul>
</div>
<div class="main-inner">
    <article class="post" itemscope itemtype="http://schema.org/BlogPosting">
        <div class="post-header">
            <div class="post-author clearfix">
                <a class="avatar fleft" href="https://github.com/Landau1994"
                   target="_blank">
                    <img width="48" src="/images/faceicon.png" alt="avatar"/>
                </a>
                <p><span class="label">作者</span>
                    <a href="https://github.com/Landau1994"
                       target="_blank">Landau1994</a>
                    <span title="最后编辑于&nbsp;2020-04-12">2020-04-12</span>
                </p>
                <p>A PhD student in bioinformatics</p>
            </div>
            <h2 class="post-title">LearnSeurat_CITE_seq</h2>
            <div class="post-meta">
                本文共计12855个字 |
                您是第&nbsp;<span id="busuanzi_value_page_pv"><i class="fa fa-spinner fa-spin"></i></span>位看到它们的小伙伴
            </div>
        </div>
        <div class="post-content markdown-body">
            <h3 id="前言">前言</h3>
<p>CITE-seq是Rahul Satija和Peter Smibert两个组合作开发的在单细胞精度，同时测量细胞表面蛋白表达和转录组的<a target="_blank" rel="noopener" href="https://www.nature.com/articles/nmeth.4380">技术</a>。该技术原理如下：</p>
<figure>
<img src="https://imgkr.cn-bj.ufileos.com/b7e96d47-230d-4519-a758-c3aa23939b18.png" alt="CITE-seq原理图，用抗体来源标签，实现细胞表面蛋白定量"><figcaption aria-hidden="true">CITE-seq原理图，用抗体来源标签，实现细胞表面蛋白定量</figcaption>
</figure>
<p>该项技术可以用于免疫相关的单细胞测序研究中。例如, 有<a target="_blank" rel="noopener" href="https://www.nature.com/articles/s41586-020-2134-y">研究表明</a>称：</p>
<blockquote>
<p>他们在人和小鼠非小细胞肺癌中进行单细胞RNA测序，鉴定了一群DC，并将其命名为“富含免疫调节分子的成熟DC”（mregDC），这是由于它们共表达了免疫调节基因（Cd274，Pdcd1lg2和Cd200）和成熟基因（Cd40，Ccr7和Il12b）。</p>
</blockquote>
<p>这段中文报道来自<a target="_blank" rel="noopener" href="http://news.sciencenet.cn/htmlpaper/2020/3/20203301537131655622.shtm">小柯机器人</a></p>
<p>Rahul Satija组开发的软件Seurat有一个<a target="_blank" rel="noopener" href="https://satijalab.org/seurat/v3.1/multimodal_vignette.html">教程</a>，可以分析CITE-seq数据。本文基于该教程对该类型数据的分析进行说明。</p>
<h3 id="数据载入">数据载入</h3>
<p>首先我们需要获取数据，该数据集取样为8617个脐带血单核细胞，包含了表达谱数据和11个抗体来源标签数据（antibody-derived tags ,ADT)。</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="do">### AvailableData() check avaliable data: we choose cbmc</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="do">### InstallData('cbmc')</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cbmc.SeuratData)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"cbmc"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="do">### expression matrix</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>cbmc[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
<pre><code>## 10 x 10 sparse Matrix of class "dgCMatrix"
##                             
## A1BG     . . . . . . . . . .
## A1BG-AS1 . . . . . . . . . .
## A1CF     . . . . . . . . . .
## A2M      . . . . . . . . . .
## A2M-AS1  . . . . . . . 1 . .
## A2ML1    . . . . . . . . . .
## A4GALT   . . . . . . . . . .
## A4GNT    . . . . . . . . . .
## AAAS     . . . . . . . . . 1
## AACS     . . . . . . . . . .</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">### ADT count matrix</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="do">### Actually there are just 10 surface protein</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>cbmc[[<span class="st">"ADT"</span>]]<span class="sc">@</span>counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>] </span></code></pre></div>
<pre><code>## 10 x 10 sparse Matrix of class "dgCMatrix"
##                                                
## CD3     60   52  89  55  63  82  53  42 103  56
## CD4     72   49 112  66  80  78  63  59 122  70
## CD8     76   59  61  56  94  57  61  55  64  80
## CD45RA 575 3943 682 378 644 479 487 472 540 535
## CD56    64   68  87  58 104  44  64  48 136  91
## CD16   161  107 117  82 168  92  77  99 235 131
## CD11c   77   65  65  44  92  63  70  75 106  69
## CD14   206  129 169 136 164 122 112 111 206 204
## CD19    70  665  79  49  81  44  60  58  61 107
## CD34   179   79  78  83 152 103  79  86 144 193</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">### show default assay</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(cbmc)</span></code></pre></div>
<pre><code>## [1] "RNA"</code></pre>
<h3 id="根据基因表达进行聚类">根据基因表达进行聚类</h3>
<p>注意在默认参数的情况下，下述操作时对<code>Default Assay</code>进行的</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># standard log-normalization</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(cbmc)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># choose ~1k variable features</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(cbmc)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># standard scaling (no regression)</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cbmc)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Run PCA, select 13 PCs for tSNE visualization and graph-based clustering</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(cbmc, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>下面的图是根据标准差来选择PCs</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(cbmc, <span class="at">ndims =</span> <span class="dv">50</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-3-1.png" style="display: block; margin: auto;"></p>
<p>聚类和t-SNE降维</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(cbmc, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(cbmc, <span class="at">resolution =</span> <span class="fl">0.8</span>)</span></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 8617
## Number of edges: 347548
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8592
## Number of communities: 19
## Elapsed time: 3 seconds</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">RunTSNE</span>(cbmc, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">25</span>, <span class="at">method =</span> <span class="st">"FIt-SNE"</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Find the markers that define each cluster, and use these to annotate the clusters, we use</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># max.cells.per.ident to speed up the process</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>cbmc.rna.markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(cbmc, <span class="at">max.cells.per.ident =</span> <span class="dv">100</span>, <span class="at">min.diff.pct =</span> <span class="fl">0.3</span>, <span class="at">only.pos =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Note, for simplicity we are merging two CD14+ Monocyte clusters (that differ in expression of</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co"># HLA-DR genes) and NK clusters (that differ in cell cycle stage)</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Memory CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Naive CD4 T"</span>, <span class="st">"NK"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Mouse"</span>, <span class="st">"B"</span>, </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CD8 T"</span>, <span class="st">"CD16+ Mono"</span>, <span class="st">"T/Mono doublets"</span>, <span class="st">"NK"</span>, <span class="st">"CD34+"</span>, <span class="st">"Multiplets"</span>, <span class="st">"Mouse"</span>, <span class="st">"Eryth"</span>, <span class="st">"Mk"</span>, </span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Mouse"</span>, <span class="st">"DC"</span>, <span class="st">"pDCs"</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(new.cluster.ids) <span class="ot">&lt;-</span> <span class="fu">levels</span>(cbmc)</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(cbmc, new.cluster.ids)</span></code></pre></div>
<p>我们看看聚类结果：</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(cbmc, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code></pre></div>
<pre><code>## Warning: Using `as.character()` on a quosure is deprecated as of rlang 0.3.0.
## Please use `as_label()` or `as_name()` instead.
## This warning is displayed once per session.</code></pre>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-5-1.png" style="display: block; margin: auto;"></p>
<h3 id="蛋白表达数据处理">蛋白表达数据处理</h3>
<p>Seurat3的assay实现多个组学或者模态的数据的存储和获取。 代码里的注释来自Seurat官网。</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we can repeat the preprocessing (normalization and scaling) steps that we typically run</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># with RNA, but modifying the 'assay' argument.  For CITE-seq data, we do not recommend typical</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co"># LogNormalization. Instead, we use a centered log-ratio (CLR) normalization, computed</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># independently for each feature.  This is a slightly improved procedure from the original</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># publication, and we will release more advanced versions of CITE-seq normalizations soon.</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(cbmc, <span class="at">assay =</span> <span class="st">"ADT"</span>, <span class="at">normalization.method =</span> <span class="st">"CLR"</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(cbmc, <span class="at">assay =</span> <span class="st">"ADT"</span>)</span></code></pre></div>
<p>在RNA表达谱的降维Embedding中同时展示展示蛋白表达水平和基因表达水平：</p>
<p>散点图：横纵轴为降维的坐标：</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># in this plot, protein (ADT) levels are on top, and RNA levels are on the bottom</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(cbmc, </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"adt_CD3"</span>, <span class="st">"adt_CD11c"</span>, </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"adt_CD8"</span>, <span class="st">"adt_CD16"</span>, </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"CD3E"</span>, <span class="st">"ITGAX"</span>, <span class="st">"CD8A"</span>, <span class="st">"FCGR3A"</span>),</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">"q05"</span>, </span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">max.cutoff =</span> <span class="st">"q95"</span>, </span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-7-1.png" style="display: block; margin: auto;"></p>
<p>Ridge Plot:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">RidgePlot</span>(cbmc, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"adt_CD3"</span>, <span class="st">"adt_CD8"</span>, <span class="st">"CD3E"</span>,<span class="st">"CD8A"</span>),<span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-8-1.png" style="display: block; margin: auto;"></p>
<p>散点图：横纵轴为表达量；这个类似于FACS</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Draw ADT scatter plots (like biaxial plots for FACS). Note that you can even 'gate' cells if</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># desired by using HoverLocator and FeatureLocator</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(cbmc, <span class="at">feature1 =</span> <span class="st">"adt_CD19"</span>, <span class="at">feature2 =</span> <span class="st">"adt_CD3"</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-9-1.png" style="display: block; margin: auto;"></p>
<p>我们也可以看看蛋白表达和基因表达的关系：</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># view relationship between protein and RNA</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(cbmc, <span class="at">feature1 =</span> <span class="st">"adt_CD3"</span>, <span class="at">feature2 =</span> <span class="st">"CD3E"</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-10-1.png" style="display: block; margin: auto;"></p>
<p>我们可以看看T细胞：</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's plot CD4 vs CD8 levels in T cells</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>tcells <span class="ot">&lt;-</span> <span class="fu">subset</span>(cbmc, <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"Naive CD4 T"</span>, <span class="st">"Memory CD4 T"</span>, <span class="st">"CD8 T"</span>))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(tcells, <span class="at">feature1 =</span> <span class="st">"adt_CD4"</span>, <span class="at">feature2 =</span> <span class="st">"adt_CD8"</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-11-1.png" style="display: block; margin: auto;"></p>
<p>选没有标准化的原始数据我们看看，坐标轴的间距太大，会有misleading</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # Let's look at the raw (non-normalized) ADT counts. You can see the values are quite high,</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># particularly in comparison to RNA values. This is due to the significantly higher protein copy</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># number in cells, which significantly reduces 'drop-out' in ADT data</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(tcells, <span class="at">feature1 =</span> <span class="st">"adt_CD4"</span>, <span class="at">feature2 =</span> <span class="st">"adt_CD8"</span>, <span class="at">slot =</span> <span class="st">"counts"</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-12-1.png" style="display: block; margin: auto;"></p>
<p>这里还是可以观察到dropouts现象的，据原作者说： &gt; If you look a bit more closely, you’ll see that our CD8 T cell cluster is enriched for CD8 T cells, but still contains many CD4+ CD8- T cells. This is because Naive CD4 and CD8 T cells are quite similar transcriptomically, and the RNA dropout levels for CD4 and CD8 are quite high. This demonstrates the challenge of defining subtle immune cell differences from scRNA-seq data alone.</p>
<p>画热图，Seurat3 加了 downsample的功能。</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Downsample the clusters to a maximum of 300 cells each (makes the heatmap easier to see for small clusters)</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>cbmc.small <span class="ot">&lt;-</span> <span class="fu">subset</span>(cbmc, <span class="at">downsample =</span> <span class="dv">300</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Find protein markers for all clusters, and draw a heatmap</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>adt.markers <span class="ot">&lt;-</span> <span class="fu">rownames</span>(cbmc.small[[<span class="st">"ADT"</span>]]<span class="sc">@</span>counts)</span></code></pre></div>
<p>我们可以看看Seurat热图的默认配色（三个冒号可以看更为底层的函数）, 个人觉得并不好看。</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># using code from RColorBrewer to demo the palette</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">=</span> <span class="dv">200</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">1</span>))</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="fu">image</span>(</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span><span class="sc">:</span>n, <span class="dv">1</span>, <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">:</span>n),</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> Seurat<span class="sc">:::</span><span class="fu">PurpleAndYellow</span>(<span class="at">k=</span>n),</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"PurpleAndYellow n"</span>, <span class="at">ylab =</span> <span class="st">""</span>, <span class="at">xaxt =</span> <span class="st">"n"</span>, <span class="at">yaxt =</span> <span class="st">"n"</span>, <span class="at">bty =</span> <span class="st">"n"</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="fu">image</span>(</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span><span class="sc">:</span>n, <span class="dv">1</span>, <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">:</span>n),</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"navy"</span>, <span class="st">"white"</span>, <span class="st">"firebrick3"</span>))(n),</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"NavyWhite3Firebrick3 n"</span>, <span class="at">ylab =</span> <span class="st">""</span>, <span class="at">xaxt =</span> <span class="st">"n"</span>, <span class="at">yaxt =</span> <span class="st">"n"</span>, <span class="at">bty =</span> <span class="st">"n"</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="fu">image</span>(</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span><span class="sc">:</span>n, <span class="dv">1</span>, <span class="fu">as.matrix</span>(<span class="dv">1</span><span class="sc">:</span>n),</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">colorRampPalette</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">11</span>,<span class="st">"RdBu"</span>))(n),</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"RdBu n"</span>, <span class="at">ylab =</span> <span class="st">""</span>, <span class="at">xaxt =</span> <span class="st">"n"</span>, <span class="at">yaxt =</span> <span class="st">"n"</span>, <span class="at">bty =</span> <span class="st">"n"</span></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-14-1.png" style="display: block; margin: auto;"></p>
<p>把默认配色换掉,见</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>mypal <span class="ot">&lt;-</span> <span class="fu">rev</span>(<span class="fu">colorRampPalette</span>(RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">11</span>,<span class="st">"RdBu"</span>))(<span class="dv">256</span>))</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co">#mypal2 &lt;- colorRampPalette(c("navy", "white", "firebrick3"))(256)</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(cbmc.small, </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">features =</span> <span class="fu">unique</span>(adt.markers), </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>          <span class="at">assay =</span> <span class="st">"ADT"</span>, </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>          <span class="at">angle =</span> <span class="dv">90</span>,<span class="at">size =</span> <span class="dv">3</span>)<span class="sc">+</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradientn</span>(<span class="at">colors  =</span> mypal)</span></code></pre></div>
<pre><code>## Scale for 'fill' is already present. Adding another scale for 'fill', which
## will replace the existing scale.</code></pre>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-15-1.png" style="display: block; margin: auto;"></p>
<p>去除细胞杂质，</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># You can see that our unknown cells co-express both myeloid and lymphoid markers (true at the</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># RNA level as well). They are likely cell clumps (multiplets) that should be discarded. We'll</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the mouse cells now as well</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">subset</span>(cbmc, <span class="at">idents =</span> <span class="fu">c</span>(<span class="st">"Multiplets"</span>, <span class="st">"Mouse"</span>), <span class="at">invert =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<h3 id="直接根据蛋白质表达水平进行聚类">直接根据蛋白质表达水平进行聚类</h3>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Because we're going to be working with the ADT data extensively, we're going to switch the</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># default assay to the 'CITE' assay.  This will cause all functions to use ADT data by default,</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># rather than requiring us to specify it each time</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(cbmc) <span class="ot">&lt;-</span> <span class="st">"ADT"</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(cbmc, <span class="at">features =</span> <span class="fu">rownames</span>(cbmc), <span class="at">reduction.name =</span> <span class="st">"pca_adt"</span>, <span class="at">reduction.key =</span> <span class="st">"pca_adt_"</span>, </span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>再来看PCA(其实这里算是degenrate到线性组合了)</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(cbmc, <span class="at">reduction =</span> <span class="st">"pca_adt"</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-18-1.png" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Since we only have 10 markers, instead of doing PCA, we'll just use a standard euclidean</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="co"># distance matrix here.  Also, this provides a good opportunity to demonstrate how to do</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co"># visualization and clustering using a custom distance matrix in Seurat</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>adt.data <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(cbmc, <span class="at">slot =</span> <span class="st">"data"</span>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>adt.dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">t</span>(adt.data))</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Before we recluster the data on ADT levels, we'll stash the RNA cluster IDs for later</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>cbmc[[<span class="st">"rnaClusterID"</span>]] <span class="ot">&lt;-</span> <span class="fu">Idents</span>(cbmc)</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Now, we rerun tSNE using our distance matrix defined only on ADT (protein) levels.</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>cbmc[[<span class="st">"tsne_adt"</span>]] <span class="ot">&lt;-</span> <span class="fu">RunTSNE</span>(adt.dist, <span class="at">assay =</span> <span class="st">"ADT"</span>, <span class="at">reduction.key =</span> <span class="st">"adtTSNE_"</span>)</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>cbmc[[<span class="st">"adt_snn"</span>]] <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(adt.dist)<span class="sc">$</span>snn</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(cbmc, <span class="at">resolution =</span> <span class="fl">0.2</span>, <span class="at">graph.name =</span> <span class="st">"adt_snn"</span>)</span></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 7895
## Number of edges: 258146
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.9491
## Number of communities: 11
## Elapsed time: 2 seconds</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can compare the RNA and protein clustering, and use this to annotate the protein clustering</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (we could also of course use FindMarkers)</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>clustering.table <span class="ot">&lt;-</span> <span class="fu">table</span>(<span class="fu">Idents</span>(cbmc), cbmc<span class="sc">$</span>rnaClusterID)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>clustering.table</span></code></pre></div>
<pre><code>##     
##      Memory CD4 T CD14+ Mono Naive CD4 T   NK    B CD8 T CD16+ Mono
##   0          1754          0        1217   29    0    27          0
##   1             0       2189           0    4    0     0         30
##   2             3          0           2  890    3     1          0
##   3             0          4           0    2  319     0          2
##   4            24          0          18    4    1   243          0
##   5             1         27           4  157    2     2         10
##   6             4          5           0    1    0     0          0
##   7             4         59           4    0    0     0          9
##   8             0          9           0    2    0     0        179
##   9             0          0           1    0    0     0          0
##   10            1          0           2    0   25     0          0
##     
##      T/Mono doublets CD34+ Eryth   Mk   DC pDCs
##   0                5     2     4   24    1    2
##   1                1     1     5   25   55    0
##   2                0     1     3    7    2    1
##   3                0     2     2    3    0    0
##   4                0     0     1    2    0    0
##   5               56     0     9   16    6    2
##   6                1   113    81   16    5    0
##   7              117     0     0    2    0    1
##   8                0     0     0    1    0    0
##   9                0     0     0    0    1   43
##   10               2     0     0    0    0    0</code></pre>
<p>下面这个embeding 还是根据ADT来的（不过只要marker连续，只有10个也没有关系？）</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"NK"</span>, <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, <span class="st">"NK"</span>, <span class="st">"CD34+"</span>, <span class="st">"T/Mono doublets"</span>, </span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CD16+ Mono"</span>, <span class="st">"pDCs"</span>, <span class="st">"B"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(new.cluster.ids) <span class="ot">&lt;-</span> <span class="fu">levels</span>(cbmc)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>cbmc <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(cbmc, new.cluster.ids)</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>tsne_rnaClusters <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(cbmc, <span class="at">reduction =</span> <span class="st">"tsne_adt"</span>, <span class="at">group.by =</span> <span class="st">"rnaClusterID"</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>tsne_rnaClusters <span class="ot">&lt;-</span> tsne_rnaClusters <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Clustering based on scRNA-seq"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>tsne_rnaClusters <span class="ot">&lt;-</span> <span class="fu">LabelClusters</span>(<span class="at">plot =</span> tsne_rnaClusters, <span class="at">id =</span> <span class="st">"rnaClusterID"</span>, <span class="at">size =</span> <span class="dv">4</span>)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>tsne_adtClusters <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(cbmc, <span class="at">reduction =</span> <span class="st">"tsne_adt"</span>, <span class="at">pt.size =</span> <span class="fl">0.5</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>tsne_adtClusters <span class="ot">&lt;-</span> tsne_adtClusters <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Clustering based on ADT signal"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>tsne_adtClusters <span class="ot">&lt;-</span> <span class="fu">LabelClusters</span>(<span class="at">plot =</span> tsne_adtClusters, <span class="at">id =</span> <span class="st">"ident"</span>, <span class="at">size =</span> <span class="dv">4</span>)</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Note: for this comparison, both the RNA and protein clustering are visualized on a tSNE</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a><span class="co"># generated using the ADT distance matrix.</span></span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a><span class="fu">wrap_plots</span>(<span class="fu">list</span>(tsne_rnaClusters, tsne_adtClusters), <span class="at">ncol =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="/figure/posts/LearnSeurat_CITEseq_files/figure-gfm/unnamed-chunk-20-1.png" style="display: block; margin: auto;"> 对于该结果，作者是这么解释的：</p>
<blockquote>
<p>The ADT-based clustering yields similar results, but with a few differences + Clustering is improved for CD4/CD8 T cell populations, based on the robust ADT data for + CD4, CD8, CD14, and CD45RA + However, some clusters for which the ADT data does not contain good distinguishing protein markers (i.e. Mk/Ery/DC) lose separation You can verify this using FindMarkers at the RNA level, as well</p>
</blockquote>
<h3 id="更多">更多</h3>
<p>pbmc 10k的细胞也提供了CITE-seq的多模态数据，具体细节，请看Seurat官方<a target="_blank" rel="noopener" href="https://satijalab.org/seurat/v3.1/multimodal_vignette.html">教程</a>。</p>
<svg style="display: none" id="MJX-SVG-global-cache"><defs></defs></svg>
            
                

            
        </div>
        <div class="post-tool">
            <a class="btn-thumbs-up" href="javascript:void(0);" data-cid="52" title="95">
                <i class="fa fa-thumbs-up" aria-hidden="true"></i> 打赏
            </a>
        </div>
        
        <div class="post-tags">标签：
            
            <a href="/tags/R/">R</a>
            
            <a href="/tags/scRNA-seq/">scRNA-seq</a>
            
            <a href="/tags/sc-seq/">sc-seq</a>
            
        </div>
        
    </article>
    
        <p style="text-align: center">本文代表个人观点，内容仅供参考。若有不恰当之处，望不吝赐教！</p>
    
    
    

</div>

<script src="/js/busuanzi.pure.mini.js"></script>



        </div><!-- end #main-->
    </div><!-- end #body -->
    <footer class="footer">
    <div class="footer-inner" style="text-align: center">
        <p>
            <a href="/about"  title="关于">关于</a>&nbsp;&nbsp<em>·</em>&nbsp;&nbsp
            <!-- 自定义链接 -->
            <a href="/help" title="帮助" >帮助</a>&nbsp;&nbsp<em>·</em>&nbsp;&nbsp
            <a href="/links" title="友链">友链</a>&nbsp;&nbsp<em>·</em>&nbsp;&nbsp
            <a href="/sitemap.xml" title="地图">地图</a>
        </p>
        <p>
            本站已建立&nbsp<a href="/timeline" id="siteBuildingTime"></a>&nbsp天，<a href="https://creativecommons.org/licenses/by-sa/4.0/" target="_blank" rel="licence">采用知识共享署名-相同方式共享 4.0 国际许可协议创作</a><br/>
            ©2017-<span id="cpYear"></span> 基于&nbsp<a href="http://hexo.io" target="_blank" rel="nofollow">Hexo</a>
            ，主题采用&nbsp&nbsp<a href="https://github.com/tangkunyin/hexo-theme-jsimple" target="_blank" rel="bookmark">JSimple</a>
            ，作者&nbsp<a href="https://landau1994.github.io" target="_blank" rel="friend">landau1994</a>
            ，Hosted by <a href="https://pages.github.com/" target="_blank" rel="nofollow">GitHub Pages</a>
        </p>
    </div>
</footer>

<script src="/js/SimpleCore.js"></script>


</div>
<!-- search pop -->
<div class="popup search-popup local-search-popup">
    <div class="local-search-header clearfix">
        <span class="search-icon">
            <i class="fa fa-search"></i>
        </span>
        <span class="popup-btn-close">
            <i class="fa fa-times-circle"></i>
        </span>
        <div class="local-search-input-wrapper">
            <input id="local-search-input"
                   spellcheck="false"
                   type="text"
                   autocomplete="off"
                   placeholder="请输入查询关键词"/>
        </div>
    </div>
    <div id="local-search-result"></div>
</div>
<div class="fixed-btn">
    <a class="btn-gotop" href="javascript:"> <i class="fa fa-angle-up"></i></a>
</div>
<script>
    $(function () {
        var jsi_config = {
            buildingTime: '08/01/2018',
            current: $('.post-tags').length > 0 ? 'post' : 'archive',
            snsQRCode: '/images/sns-qrcode.png',
            donateImg: '/images/donate-qr.png',
            localSearch: { dbPath: '' },
            readMode: 'day'
        };
        
            jsi_config.localSearch = {
                dbPath: '/search.json',
                trigger: 'auto',
                topN: '1',
                unescape: 'false'
            }
        
        SimpleCore.init(jsi_config);
        
    });
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ["$","$"], ["\\(","\\)"] ],
            skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code'],
            processEscapes: true
        }
    });
    MathJax.Hub.Queue(function() {
        var all = MathJax.Hub.getAllJax();
        for (var i = 0; i < all.length; ++i)
            all[i].SourceElement().parentNode.className += ' has-jax';
    });
</script>
<script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
</body>
</html>
